AWS SpotFleet support - very basic

fprochazka opened the pull request:
I know this repo is in maintenance - this is here for other users that might want to pick this up. If this interests the maintainers of this repo, I'd be willing to invest some more time to make this mergeable (this was my first encounter with Go).

------

**Problem:** Basic aws spot instances cannot automatically (to my knowledge) use different availability zone, based on what spot pools are depleted.

**Solution:** Introduce _basic_ support for AWS SpotFleet. This is a minimal implementation, that solves the problem _for me_, but anyone can use it as inspiration or directly.

I will **not** be supporting this fork. Use at your own risk. I will maintain the fork only for my own needs.

## Compiling my fork:

* setup the repo like the [Contributing guide says](https://github.com/docker/machine/blob/61ef47dc5d6b1658e3d6636f9382d50507c8c7e1/CONTRIBUTING.md#local-go-development-environment)
* `cd src/github.com/docker/machine`
* make sure you know how to compile the main repo, before continuing
* `git remote add fprochazka git@github.com:fprochazka/docker-machine.git`
* `git fetch fprochazka`
* `git checkout fp-allow-multiple-availability-zones`
* `make build`

I have build & tested this with 

* `go version go1.12.4 linux/amd64`
* `dep version devel` (installed today)

## Usage: basic

I'm using the `docker-machine` for my [gitlab runners](https://docs.gitlab.com/runner/configuration/runner_autoscale_aws/index.html). 

1. test in EC2 Console, that you're able to request and cancel SpotFleet (this should initialize the required role, if your account doesn't already have it)
2. find out your `vpc-id` and all the `subnet-id`'s of availability zones, that you want to use.
3. find out your "IAM fleet role" (the EC2 SpotFleet dialog shows it to you)
4. profit

```sh
docker-machine create --driver amazonec2 \
      --amazonec2-access-key "AKIA" \
      --amazonec2-secret-key "secret" \
      --amazonec2-region "eu-central-1" \
      --amazonec2-vpc-id "vpc-abc123" \
      --amazonec2-subnet-id "subnet-abc1" \
      --amazonec2-subnet-id "subnet-abc2" \
      --amazonec2-subnet-id "subnet-abc3" \
      --amazonec2-instance-type "t3.small" \
      --amazonec2-tags "Name,Gitlab CI Runner,user:Application,Gitlab,user:Stack,production,gitlab,true,gitlab-runner-autoscale,true" \
      --amazonec2-security-group "gitlab-runners" \
      --amazonec2-security-group-readonly \
      --amazonec2-request-spot-fleet \
      --amazonec2-iam-fleet-role "arn:aws:iam::123:role/aws-ec2-spot-fleet-tagging-role" \
      --amazonec2-use-private-address \
      spot-fleet-test
```

## Usage: with gitlab-runner

```toml
  [runners.machine]
    IdleCount = 0
    IdleTime = 1800
    OffPeakIdleCount = 0
    OffPeakIdleTime = 0
    MachineDriver = "amazonec2"
    MachineName = "gitlab-runner-micro-%s"
    MachineOptions = [
      "amazonec2-access-key=AKIA",
      "amazonec2-secret-key=secret",
      "amazonec2-instance-type=t3.micro",
      "amazonec2-tags=Name,Gitlab CI Runner,user:Application,Gitlab,user:Stack,production,gitlab,true,gitlab-runner-autoscale,true",
      "amazonec2-region=eu-central-1", 
      "amazonec2-vpc-id=vpc-abc123", 
      "amazonec2-subnet-id=subnet-abc1", 
      "amazonec2-subnet-id=subnet-abc2",
      "amazonec2-subnet-id=subnet-abc3",
      "amazonec2-use-private-address=true",
      "amazonec2-security-group=gitlab-runners", 
      "amazonec2-security-group-readonly=true",
      "amazonec2-request-spot-fleet=true",
      "amazonec2-iam-fleet-role=arn:aws:iam::123:role/aws-ec2-spot-fleet-tagging-role"
    ]
```

ayufan committed:
Make DigitalOcean driver RateLimit aware

Signed-off-by: Tomasz Maczukin <tomasz@maczukin.pl>

ayufan committed:
Save machine when create fails

Signed-off-by: Tomasz Maczukin <tomasz@maczukin.pl>

tmaczukin committed:
Add backoff mechanism to google driver operations checks

Signed-off-by: Tomasz Maczukin <tomasz@maczukin.pl>

tmaczukin committed:
Update build configuration

tmaczukin committed:
Add --google-min-cpu-platform option for machine creation

Signed-off-by: Tomasz Maczukin <tomasz@maczukin.pl>

fprochazka committed:
support for AWS SpotFleet

Signed-off-by: Filip Prochazka <filip@prochazka.su>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "fp-allow-multiple-availability-zones" git@github.com:fprochazka/docker-machine.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842359068536
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

declension commented:
Any progress on this? It'd be great to have Spot Fleet support!

fprochazka commented:
@declension what kind of progress do you expect? :) The maintainers already said no more features.

declension commented:
> The maintainers already said no more features.

Oh :disappointed:  Can't find any evidence of this on the home page or contributing guides though, and haven't followed.

Could this become a [plugin](https://github.com/docker/docker.github.io/blob/master/machine/AVAILABLE_DRIVER_PLUGINS.md) or even fork, somehow?

fprochazka commented:
@declension https://github.com/docker/machine/issues/4537

You actually can use my fork with this feature very easily, I've tried to describe it in the description. I'm using it on our production. But at your own risk ;-)

mohag commented:
It might be useful to get into rancher/machine, which is used by Rancher (and seems to be getting more features than this...)

Commit:
Bump to v0.16.2-gitlab.1

Sync branch with upstream https://github.com/docker/machine/

Commit:
Create new image for CI

Create a new image to be used for CI, this will be used for testing,
building and uploading to s3.

Put the extra files inside of the `.gitlab` directory to prevent any
conflicts since this is a fork of a project.

Commit:
Add merge request pipelines

Dogfood merge request pipelines to this small project.

Commit:
Release binaries on S3 bucket

Create a new release job to release the binaries for each commit,
master, and tag created into their respective bucket.

The s3 bucket information are stored inside of **Settings > CI/CD >
Variables**

tmaczukin committed:
Fix version pattern in version.sh

tmaczukin committed:
Fix beta release refs definition

Commit:
Use cached IP Address for amazonec2 driver

`GetIP` gets called every time docker-machine wants to send an ssh
command. For the amazonec2 driver it is not using the saved IPAddress
but calling the aws API every time which leads to an excessive amount of
IP calls which exhaust the API limit. Before sending a request for the
IP check if it's populated, the IPAddress is being populated for the
first time when the machine starts in
[instanceIpAvailable](https://gitlab.com/gitlab-org/ci-cd/docker-machine/blob/8bf0b23c8e88e14ed6cf9221b747e1ce062fc6ba/drivers/amazonec2/amazonec2.go#L543)

Reading [amazon
documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-instance-addressing.html#concepts-public-addresses) on weather a public IP can change during
the lifecycle of a machine it does not seem like it will effect
docker-machine in any way.

This seems to effect other drivers like the `google` driver, but drivers
like `digitalocean` are not effected since the use the
[`baseDriver.GetIP`](https://gitlab.com/gitlab-org/ci-cd/docker-machine/blob/8bf0b23c8e88e14ed6cf9221b747e1ce062fc6ba/libmachine/drivers/base.go#L39-44)
which does not call the API. Also on drivers like
[softlayer](https://gitlab.com/gitlab-org/ci-cd/docker-machine/blob/8bf0b23c8e88e14ed6cf9221b747e1ce062fc6ba/drivers/softlayer/driver.go#L285-302)
it is doing the same check before calling the API.

We can see significant reduction of API calls:

**Before Patch:**
Number of DescribeInstances: 34
Number of GetSSHHostname: 17

**After Patch:**
Number of DescribeInstances: 13
Number of GetSSHHostname: 17

More investigation of this issue can be found in https://gitlab.com/gitlab-org/gitlab-runner/issues/3424#note_231357594

reference https://gitlab.com/gitlab-org/gitlab-runner/issues/3424

Commit:
Use cached IP for google driver

`GetIP` gets called every time `GetSSHHostname` is called which is
called to run every ssh command to the machine. `GetIP` is sending a
request to the API every time we request it.

Check if the in-memory IP is available before sending a request to the
API, and save it at the end as well if it's not present.

Adding tests is quite difficult here since the client for accessing the
Google API is created within `GetIP` so it's not something that can be
mocked easily. To make it testable we require a lot of changes to the
google driver which will be riskier the not adding tests.

Below are some metrics on how many times we used to call the API for the
IP address when creating and removing a machine:

**Before Patch:**
Number of `GetIP` Calls: 21
Number of `Instance.Get` Calls: 21

**After Patch:**
Number of `GetIP` Calls: 21
Number of `Instance.Get` Calls: 1

tmaczukin committed:
Fix the project title added to S3 release index file

tmaczukin committed:
Bump version to 0.16.2-gitlab.3

Commit:
Add warning about contributing to GitLab's fork

Commit:
Add link to docs

tmaczukin committed:
Switch to release-index-gen for S3 index.html generation

Commit:
Remove dependency on `gitlab-runner-builder` runner

Migrate the following variables to **Settings > CI/CD > Variables**

- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`

reference https://gitlab.com/gitlab-org/gitlab-runner/-/issues/21299

tmaczukin committed:
Pin CI jobs to gitlab-org runners

resmo committed:
provision: fix flatcar detection in coreos

CoreOS is deprecated and Flatcar is the successor. This simple addition
would fix the detecttion.

Flatcar has an ID_LIKE key with value coreos:

core@localhost ~ $ cat /etc/os-release
NAME="Flatcar Container Linux by Kinvolk"
ID=flatcar
ID_LIKE=coreos
VERSION=2345.3.1
VERSION_ID=2345.3.1
BUILD_ID=2020-03-26-2026
PRETTY_NAME="Flatcar Container Linux by Kinvolk 2345.3.1 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://flatcar-linux.org/"
BUG_REPORT_URL="https://issues.flatcar-linux.org"
FLATCAR_BOARD="amd64-usr"

A test output looks as expected:

$ ./bin/docker-machine create --driver cloudscale --cloudscale-flavor flex-2 --cloudscale-image flatcar-2303.3 --cloudscale-zone lpg1 --cloudscale-ssh-user core --cloudscale-token=$CLOUDSCALE_API_TOKEN flatcar
Running pre-create checks...
Creating machine...
(flatcar2) Creating SSH key...
(flatcar2) Starting to create cloudscale.ch server...
(flatcar2) Waiting for IP address to be assigned to the Server...
Waiting for machine to be running, this may take a few minutes...
Detecting operating system of created instance...
Waiting for SSH to be available...
Detecting the provisioner...
Provisioning with coreOS...
Copying certs to the local machine directory...
Copying certs to the remote machine...
Setting Docker configuration on the remote daemon...
Checking connection to Docker...
Docker is up and running!

Signed-off-by: Rene Moser <mail@renemoser.net>

tmaczukin committed:
Handle the error that google's firewallRule() can return

tmaczukin committed:
Bump version to 0.16.2-gitlab.4

Commit:
Handle case where no instances are returned after calling AWS RunInstances

tmaczukin committed:
Run development release only for gitlab-org changes

Commit:
Add fork warning to readme

ghost commented:
I tried this on gitlab-runner v12.9.1. but dose not work.

- spot fleet request was created
![image](https://user-images.githubusercontent.com/6925571/84731346-8f93a680-afd3-11ea-8848-9675253f0f74.png)

- But Docker machine provisioning fails on SSH connection

- log text

```
Checking for jobs... received                       job=8766 repo_url=http://xxxx/xxxx.git runner=bdc652f8
Creating CA: /root/.docker/machine/certs/ca.pem     driver=amazonec2 name=runner-bdc652f8-runner-spottest-1592280235-612023db operation=create
Creating client certificate: /root/.docker/machine/certs/cert.pem  driver=amazonec2 name=runner-bdc652f8-runner-spottest-1592280235-612023db operation=create
Running pre-create checks...                        driver=amazonec2 name=runner-bdc652f8-runner-spottest-1592280235-612023db operation=create
Creating machine...                                 driver=amazonec2 name=runner-bdc652f8-runner-spottest-1592280235-612023db operation=create
(runner-bdc652f8-runner-spottest-1592280235-612023db) Launching instance...  driver=amazonec2 name=runner-bdc652f8-runner-spottest-1592280235-612023db operation=create
(runner-bdc652f8-runner-spottest-1592280235-612023db) Created spot fleet request sfr-a64862c1-dc27-44dc-a2ee-153cbb5e9598 ...  driver=amazonec2 name=runner-bdc652f8-runner-spottest-1592280235-612023db operation=create
WARNING: Failed to parse "X-GitLab-Trace-Update-Interval" header  error=strconv.Atoi: parsing "": invalid syntax header-value= job=8766 runner=bdc652f8
```

If you know anything, please

fprochazka commented:
@kubokkey I have the same warnings as you, but my runners are working fine, sorry I have no idea how to fix this... 

Although, I was considering forking the Gitlab-maintained docker-machine and adding my changes there, because last time I've tried rebasing this and updating dependencies my runners stopped working and I had to revert the binary.

Commit:
Also tag AWS EC2 EBS volumes

Use the input from --amazonec2-tags to also tag EBS volumes

Resolves https://github.com/docker/machine/issues/4678

Commit:
Apply review suggestions

tmaczukin committed:
Bump version to 0.16.2-gitlab.5

adrienkohlbecker committed:
Add support for instance labels on GCP

adrienkohlbecker committed:
Apply 2 suggestion(s) to 2 file(s)

adrienkohlbecker committed:
Use StringSliceFlag

adrienkohlbecker committed:
Fix spelling

krzysztof-miemiec commented:
If you came here because you'd like to find a relatively simple way to create an instance using spot fleet, check out custom [docker-machine-driver-terraform](https://github.com/krzysztof-miemiec/docker-machine-driver-terraform) and [usage example for Gitlab runner](https://github.com/krzysztof-miemiec/terraform-aws-gitlab-runner/commit/82b8aa12842269a5d6822ccae3cf8a0809ccaef5). You can also check out this PR: https://github.com/npalm/terraform-aws-gitlab-runner/pull/249. :wink:

adrienkohlbecker committed:
Update test

Commit:
Add clearer steps for when an MR is accepted

Commit:
Apply 1 suggestion(s) to 1 file(s)

Commit:
Document release process of docker-machine

The process is not complex but it's not very well known to anyone. This
issue template documents the whole process which gives an option to
improve and iterate over it.

Commit:
Update version to 0.16.2-gitlab.6

tmaczukin committed:
Add support for Container Optimized OS

tmaczukin committed:
Support passing custom metadata on Google instance creation

tmaczukin committed:
Support passing cloud-config user data on Google instance creation

tmaczukin committed:
Bump version to 0.16.2-gitlab.7

tmaczukin committed:
Fix typos in the release issue template

tmaczukin committed:
Build statically linked binary for linux/amd64

tmaczukin committed:
Bump version to 0.16.2-gitlab.8

tmaczukin committed:
Bump version to 0.16.2-gitlab.9

stanhu committed:
Add support for using GPUs in Google Compute Engine

Add support for configuring Google maintenance policy

This is needed since GPU-accelerated instances cannot have a MIGRATE
maintenance policy:
https://cloud.google.com/compute/docs/instances/live-migration

Relates to https://gitlab.com/gitlab-org/ci-cd/docker-machine/-/issues/34

Commit:
Bump version of aws-sdk-go to 1.36.21

Commit:
Run instances with IMDSv2

Commit:
Bump version to 0.16.2-gitlab.10

Commit:
Update release checklist

balluthemallu committed:
Update region.go for HK support (ap-east-1)
Signed-off-by: Arjun Balraj <arjunbalraj@email.com>

tmaczukin committed:
Bump version to 0.16.2-gitlab.11

tmaczukin committed:
Update google.golang.org/api to a newest version

igorwwwwwwwwwwwwwwwwwwww committed:
Include symbols in docker-machine binary

tmaczukin committed:
Update github.com/golang/protobuf

tmaczukin committed:
Update golang.org/x/sys

tmaczukin committed:
Update golang.org/x/net

tmaczukin committed:
Update golang.org/x/oauth2

tmaczukin committed:
Use Wait() not Get() to check operation status

tmaczukin committed:
Bump version to 0.16.2-gitlab.12

Commit:
refactor: change default branch to main

Update the default branch to be `main` instead of `master`.

reference https://gitlab.com/gitlab-org/gitlab-runner/-/issues/27239

pedropombeiro committed:
Use CI_JOB_TOKEN to create releases

Commit:
add --google-skip-firewall-create

Commit:
Bump version to 0.16.2-gitlab.13

zyga committed:
Fix nanosecond wait durations in digitalocean driver

The code meant to sleep for 2-8 seconds, ended up using nanoseconds
instead. This fixes machine allocation errors on Digital Ocean.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

darren17082 committed:
install-machine.md file to repo
